stunner/stunner.go

```diff
@@ -105,95 +110,114 @@ func (s *Stunner) Receive(p []byte, fromAddr *net.UDPAddr) {
 	}
 	r, ok := s.removeTX(tx)
 	if !ok {
-		s.logf("stunner: got STUN packet for unknown TxID %x", tx)
 		return
 	}
 	d := now.Sub(r.sent)
 
-	session := s.sessions[r.server]
-	if session != nil {
-		host := net.JoinHostPort(net.IP(addr).String(), fmt.Sprint(port))
-		s.Endpoint(r.server, host, d)
-		session.cancel()
-	}
+	host := net.JoinHostPort(net.IP(addr).String(), fmt.Sprint(port))
+	s.onPacket(r.server, host, d)
 }
 
 func (s *Stunner) resolver() *net.Resolver {
 	return net.DefaultResolver
 }
 
+// cleanUpPostRun zeros out some fields, mostly for debugging (so
+// things crash or race+fail if there's a sender still running.)
+func (s *Stunner) cleanUpPostRun() {
+	s.mu.Lock()
+	s.inFlight = nil
+	s.mu.Unlock()
+	s.onPacket = nil
```

Comment:
Doesn't this also need the lock?

Reply:
It's owned by Run in its main goroutine, which sets it on entry and clears it on exit.

Nothing else can call it concurrently once inFlight has been zeroed. I'll add a comment.

I'm always reluctant to "over lock" things because it get confusing to readers in the future. I'd rather have it be as tight as possible and catch mistakes if assumptions were wrong.

Reply:
Actually I'm just going to delete the onPacket clear. It was mostly for debugging. It's not required for correctness. The GC will deal with it.